wagtail/search/models.py

```diff
@@ -63,14 +84,81 @@ def garbage_collect(cls, days=None):
         min_date = timezone.now().date() - datetime.timedelta(days)
 
         cls.objects.filter(date__lt=min_date).delete()
 
     class Meta:
-        unique_together = (
-            ('query', 'date'),
+        unique_together = (("query", "date"),)
+        verbose_name = _("Query Daily Hits")
+        verbose_name_plural = _("Query Daily Hits")
+
+
+class TextIDGenericRelation(GenericRelation):
+    auto_created = True
+
+    def get_content_type_lookup(self, alias, remote_alias):
+        field = self.remote_field.model._meta.get_field(self.content_type_field_name)
+        return field.get_lookup("in")(
+            field.get_col(remote_alias), get_descendants_content_types_pks(self.model)
+        )
+
+    def get_object_id_lookup(self, alias, remote_alias):
+        from_field = self.remote_field.model._meta.get_field(self.object_id_field_name)
+        to_field = self.model._meta.pk
+        return from_field.get_lookup("exact")(
+            from_field.get_col(remote_alias), Cast(to_field.get_col(alias), from_field)
         )
-        verbose_name = _('Query Daily Hits')
-        verbose_name_plural = _('Query Daily Hits')
+
+    def get_extra_restriction(self, where_class, alias, remote_alias):
+        cond = where_class()
+        cond.add(self.get_content_type_lookup(alias, remote_alias), "AND")
+        cond.add(self.get_object_id_lookup(alias, remote_alias), "AND")
+        return cond
+
+    def resolve_related_fields(self):
+        return []
+
+
+class BaseIndexEntry(models.Model):
+    """
+    This is an abstract class that only contains fields common to all database vendors.
+    It should be extended by the models specific for each backend.
+    """
+
+    content_type = models.ForeignKey(
+        ContentType, on_delete=models.CASCADE, related_name="search_content_type"
+    )
+    # We do not use an IntegerField since primary keys are not always integers.
+    object_id = models.TextField()
+    content_object = GenericForeignKey()
+
+    # TODO: Add per-object boosting.
+    # This field stores the "Title Normalisation Factor"
+    # This factor is multiplied onto the the rank of the title field.
+    # This allows us to apply a boost to results with shorter titles
+    # elevating more specific matches to the top.
+    title_norm = models.FloatField(default=1.0)
+
+    class Meta:
+        unique_together = ("content_type", "object_id")
+        verbose_name = _("index entry")
+        verbose_name_plural = _("index entries")
+        abstract = True
+
+    def __str__(self):
+        return "%s: %s" % (self.content_type.name, self.content_object)
+
+    @property
+    def model(self):
+        return self.content_type.model
+
+    # TODO: this needs to be changed --decide later a new name
+    @classmethod
+    def add_generic_relations(cls):
+        for model in apps.get_models():
+            if class_is_indexed(model):
+                TextIDGenericRelation(cls).contribute_to_class(model, "index_entries")
```

Comment:
Yes, it does. I've added a check in the `ready` method of the `wagtail.search` app config, so that the generic relations to the `IndexEntry` model that will be used are not added if the `wagtail.contrib.postgres_search` app is already installed:

```python3
def ready(self):
        if not apps.is_installed('wagtail.contrib.postgres_search'):
               # Add the generic relations
```